resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala

```diff
@@ -0,0 +1,102 @@
+import org.apache.spark.storage.{BlockId, BlockManager, StorageLevel, UnrecognizedBlockId}
+import org.apache.spark.util.Utils
+
+class KubernetesLocalDiskShuffleExecutorComponents(sparkConf: SparkConf)
+  extends ShuffleExecutorComponents with Logging {
+
+  private val delegate = new LocalDiskShuffleExecutorComponents(sparkConf)
+  private var blockManager: BlockManager = _
+
+  override def initializeExecutor(
+      appId: String, execId: String, extraConfigs: java.util.Map[String, String]): Unit = {
+    delegate.initializeExecutor(appId, execId, extraConfigs)
+    blockManager = SparkEnv.get.blockManager
+    if (sparkConf.getBoolean("spark.kubernetes.driver.reusePersistentVolumeClaim", false)) {
+      // Turn off the deletion of the shuffle data in order to reuse
+      blockManager.diskBlockManager.deleteFilesOnStop = false
+      Utils.tryLogNonFatalError {
+        KubernetesLocalDiskShuffleExecutorComponents.recoverDiskStore(sparkConf, blockManager)
+      }
+    }
+  }
+
+  override def createMapOutputWriter(shuffleId: Int, mapTaskId: Long, numPartitions: Int)
+    : ShuffleMapOutputWriter = {
+    delegate.createMapOutputWriter(shuffleId, mapTaskId, numPartitions)
+  }
+
+  override def createSingleFileMapOutputWriter(shuffleId: Int, mapId: Long)
+    : Optional[SingleSpillShuffleMapOutputWriter] = {
+    delegate.createSingleFileMapOutputWriter(shuffleId, mapId)
+  }
+}
+
+object KubernetesLocalDiskShuffleExecutorComponents extends Logging {
+  /**
+   * This tries to recover shuffle data of dead executors' local dirs if exists.
+   * Since the executors are already dead, we cannot use `getHostLocalDirs`.
+   * This is enabled only when spark.kubernetes.driver.reusePersistentVolumeClaim is true.
+   */
+  def recoverDiskStore(conf: SparkConf, bm: BlockManager): Unit = {
+    // Find All files
+    val files = Utils.getConfiguredLocalDirs(conf)
+      .filter(_ != null)
+      .map(s => new File(new File(new File(s).getParent).getParent))
+      .flatMap { dir =>
+        val oldDirs = dir.listFiles().filter { f =>
+          f.isDirectory && f.getName.startsWith("spark-")
+        }
+        val files = oldDirs
+          .flatMap(_.listFiles).filter(_.isDirectory) // executor-xxx
+          .flatMap(_.listFiles).filter(_.isDirectory) // blockmgr-xxx
+          .flatMap(_.listFiles).filter(_.isDirectory) // 00
+          .flatMap(_.listFiles)
+        if (files != null) files.toSeq else Seq.empty
+      }
+
+    logInfo(s"Found ${files.size} files")
+
+    // This is not used.
+    val classTag = implicitly[ClassTag[Object]]
+    val level = StorageLevel.DISK_ONLY
+    val (indexFiles, dataFiles) = files.partition(_.getName.endsWith(".index"))
+    (dataFiles ++ indexFiles).foreach { f =>
+      try {
+        val id = BlockId(f.getName)
+        val decryptedSize = f.length()
+        bm.TempFileBasedBlockStoreUpdater(id, level, classTag, f, decryptedSize).save()
+      } catch {
+        case _: UnrecognizedBlockId =>
```

Comment:
Is it good to ignore the exception? Looks like we will lose the block?

Reply:
No, `UnrecognizedBlockId` means the file name is not in Spark-generated pattern.
```scala
@DeveloperApi
class UnrecognizedBlockId(name: String)
    extends SparkException(s"Failed to parse $name into a block ID")
```